frame/election-provider-support/src/lib.rs

```diff
@@ -214,11 +214,28 @@ pub trait ElectionDataProvider {
 	/// appropriate weight at the end of execution with the system pallet directly.
 	fn voters(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<VoterOf<Self>>>;
 
-	/// The number of targets to elect.
+	/// The uncapped number of targets to elect.
 	///
 	/// This should be implemented as a self-weighing function. The implementor should register its
 	/// appropriate weight at the end of execution with the system pallet directly.
-	fn desired_targets() -> data_provider::Result<u32>;
+	fn uncapped_desired_targets() -> data_provider::Result<u32>;
+
+	/// The number of targets to elect.
+	///
+	/// This will call `uncapped_desired_targets` which will return a value `y`.
+	/// Call `x` the size of the return of the call to `targets`, if y > x,
+	/// this function will emit a warning and return `x` instead of `y`, otherwise
+	/// return `y`. This is documented in issue: https://github.com/paritytech/substrate/issues/9478
+	fn desired_targets() -> data_provider::Result<u32> {
+		let x = Self::targets(None)?;
```

Comment:
calling `targets` is super expensive, we can't afford to call it once here, and once in `pallet-election-provider-multi-phase`. 

What I meant by the call site of `ElectionDataProvider` is in fact `pallet-election-provider-multi-phase`. 

https://github.com/paritytech/substrate/blob/42b2d623d058197aebc3c737fb44fbbf278a85b4/frame/election-provider-multi-phase/src/lib.rs#L1359